tests/unit/s2n_fingerprint_test.c

```diff
@@ -245,5 +277,440 @@ int main(int argc, char **argv)
+            EXPECT_EQUAL(raw_size, legacy_output_size);
+        };
+    };
+
+    /* Test reusing fingerprint structures */
+    {
+        DEFER_CLEANUP(struct s2n_fingerprint *fingerprint = s2n_fingerprint_new(S2N_FINGERPRINT_JA3),
+                s2n_fingerprint_free);
+
+        uint8_t raw_client_hello_2[sizeof(raw_client_hello)] = { 0 };
+        EXPECT_MEMCPY_SUCCESS(raw_client_hello_2, raw_client_hello, sizeof(raw_client_hello));
+        /* Make a minor change so that we will calculate a different fingerprint.
+         * This changes the protocol version from 1.2 to 1.3. */
+        raw_client_hello_2[5] = 0x04;
+
+        DEFER_CLEANUP(struct s2n_client_hello *client_hello_2 = NULL, s2n_client_hello_free);
+        client_hello_2 = s2n_client_hello_parse_message(raw_client_hello_2, sizeof(raw_client_hello_2));
+        EXPECT_NOT_NULL(client_hello_2);
+
+        struct s2n_test_hello {
+            struct s2n_client_hello *hello;
+            uint8_t hash[S2N_TEST_OUTPUT_SIZE];
+            uint8_t raw[S2N_TEST_OUTPUT_SIZE];
+            unsigned int init : 1;
+        } hellos[] = {
+            { .hello = client_hello },
+            { .hello = client_hello_2 },
+        };
+
+        for (size_t i = 0; i < 10; i++) {
+            struct s2n_test_hello *test_case = &hellos[i % s2n_array_len(hellos)];
+            EXPECT_SUCCESS(s2n_fingerprint_set_client_hello(fingerprint, test_case->hello));
+
+            /* Verify no state leftover after reset */
+            uint32_t size = 0;
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_fingerprint_get_raw_size(fingerprint, &size),
+                    S2N_ERR_INVALID_STATE);
+
+            /* Store initial copy of hash / raw string for later comparisons */
+            if (test_case->init == false) {
+                test_case->init = true;
+                EXPECT_SUCCESS(s2n_fingerprint_get_hash(fingerprint,
+                        sizeof(test_case->hash), test_case->hash, &size));
+                EXPECT_SUCCESS(s2n_fingerprint_get_raw(fingerprint,
+                        sizeof(test_case->raw), test_case->raw, &size));
+            }
+
+            /* Verify all hash calculations match across resets */
+            uint8_t hash_output[S2N_TEST_OUTPUT_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_fingerprint_get_hash(fingerprint,
+                    sizeof(hash_output), hash_output, &size));
+            EXPECT_BYTEARRAY_EQUAL(hash_output, test_case->hash, S2N_TEST_OUTPUT_SIZE);
+
+            /* Verify all raw string calculations match across resets */
+            uint8_t raw_output[S2N_TEST_OUTPUT_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_fingerprint_get_raw(fingerprint,
+                    sizeof(raw_output), raw_output, &size));
+            EXPECT_BYTEARRAY_EQUAL(raw_output, test_case->raw, S2N_TEST_OUTPUT_SIZE);
+
+            /* Reset for the next iteration */
+            EXPECT_SUCCESS(s2n_fingerprint_wipe(fingerprint));
```

Comment:
Is it worth testing this without manually calling wipe as well?

Reply:
I don't think so. That case is already covered by a unit test, and this is more of a functional test. The correct way to write the code is with the wipe.